servers/rendering/shader_converter.cpp

```diff
@@ -0,0 +1,2591 @@
+	return true;
+}
+
+List<SL::Token>::Element *ShaderDeprecatedConverter::replace_curr(const Token &p_token, String comment_format) {
+	ERR_FAIL_COND_V(curr_ptr == nullptr, nullptr);
+	Token new_token = p_token;
+	new_token.pos = NEW_IDENT;
+	List<Token>::Element *prev = curr_ptr;
+	if (!comment_format.is_empty()) {
+		int count = comment_format.count("%s");
+		if (count == 0) {
+			comment_format = RTR(comment_format) + " '%s' renamed to '%s'";
+		} else if (count == 1) {
+			comment_format = RTR(comment_format) + ", renamed to '%s'";
+		}
+		_add_comment_before(vformat(RTR(comment_format), get_token_literal_text(curr_ptr->get()), p_token.text), curr_ptr, false);
+	}
+	curr_ptr = code_tokens.insert_before(curr_ptr, new_token);
+	ERR_FAIL_COND_V(!code_tokens.erase(prev), nullptr);
+	return curr_ptr;
+}
+
+SL::Token ShaderDeprecatedConverter::mkTok(TokenType p_type, const StringName &p_text, double p_constant, uint16_t p_line) {
+	return { p_type, p_text, p_constant, p_line, 0, NEW_IDENT };
+}
+
+bool ShaderDeprecatedConverter::_insert_uniform_declaration(const String &p_name) {
+	if (after_shader_decl == nullptr) {
+		return false;
+	}
+	TokenType type = get_removed_builtin_uniform_type(p_name);
+	Vector<TokenType> hints = get_removed_builtin_hints(p_name);
+	Vector<Token> uni_decl = { mkTok(TT::TK_NEWLINE), mkTok(TT::TK_UNIFORM), mkTok(TT::TK_SPACE), mkTok(type),
+		mkTok(TT::TK_SPACE), mkTok(TT::TK_IDENTIFIER, p_name), mkTok(TT::TK_SPACE), mkTok(TT::TK_COLON),
+		mkTok(TT::TK_SPACE) };
+	for (int i = 0; i < hints.size(); i++) {
+		uni_decl.append(mkTok(hints[i]));
+		if (i < hints.size() - 1) {
+			uni_decl.append(mkTok(TT::TK_COMMA));
+			uni_decl.append(mkTok(TT::TK_SPACE));
+		}
+	}
+	uni_decl.append_array({ mkTok(TT::TK_SEMICOLON), mkTok(TT::TK_NEWLINE) });
+	if (!insert_after(uni_decl, after_shader_decl)) {
+		return false;
+	}
+	TokenE *cur_pos = get_pos();
+	reset_to(after_shader_decl);
+	UniformDecl uni;
+	uni.start_pos = get_next_token(); // uniform
+	uni.type_pos = get_next_token(); // type
+	uni.name_pos = get_next_token(); // id
+	get_next_token(); // colon
+	for (int i = 0; i < hints.size(); i++) {
+		uni.hint_poses.push_back(get_next_token()); // hint
+		if (i < hints.size() - 1) {
+			get_next_token(); // comma
+		}
+	}
+	uni.end_pos = get_next_token();
+	uniform_decls[p_name] = uni;
+	_add_comment_before(vformat(RTR("Usage of deprecated built-in '%s' requires a uniform declaration."), p_name), uni.start_pos, false);
+	reset_to(cur_pos);
+	return true;
+}
+
+RS::ShaderMode ShaderDeprecatedConverter::get_shader_mode_from_string(const String &p_mode) {
+	if (p_mode == "spatial") {
+		return RS::SHADER_SPATIAL;
+	} else if (p_mode == "canvas_item") {
+		return RS::SHADER_CANVAS_ITEM;
+	} else if (p_mode == "particles") {
+		return RS::SHADER_PARTICLES;
+	} else { // 3.x didn't support anything else.
+		return RS::SHADER_MAX;
+	}
+}
+// Remove from the current token to end (exclsusive) and return the new current token.
```

Comment:
```suggestion

// Remove from the current token to end (exclsusive) and return the new current token.
```
Please leave an empty line between methods